### PR TITLE
Assemble csi shared utility functions to csi_util

### LINKED ISSUE #58628: CSI Code Cleanup - Assemble shared utility functions in a central location
This is a maintenance TODO for CSI code organization.

### PR DESCRIPTION
**Pull request by Lion-Wei**
**What this PR does / why we need it**:

Assemble csi shared utility functions to csi_util and add unit test.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #58628

**Special notes for your reviewer**:

The function list moved to csi_util
```go
func log(msg string, parts ...interface{}) string{}
func getCSISourceFromSpec(spec *volume.Spec) (*v1.CSIPersistentVolumeSource, error){}
func getCSIClient(volume, driverName string) *csiDriverClient{}
func getTargetPath(uid types.UID, specVolumeID string, host volume.VolumeHost) string{}
func getAttachmentName(volName, csiDriverName, nodeName string) string{}
func saveVolumeData(p *csiPlugin, podUID types.UID, specVolID string, data map[string]string) error{}
func loadVolumeData(dir string, fileName string) (map[string]string, error){}
```
And also, extracted common part of `csiPlugin.NetAttacher` and `csiPlugin.NewDetached` to `csiPlugin.newCSIAttacher`.

**Release note**:
```release-note
NONE
```

### CODE REVIEW

File: pkg/volume/csi/csi_util.go [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+func getCSISourceFromSpec(spec *volume.Spec) (*v1.CSIPersistentVolumeSource, error) {
+	if spec.PersistentVolume != nil &&
+		spec.PersistentVolume.Spec.CSI != nil {
+		return spec.PersistentVolume.Spec.CSI, nil
+	}
+
+	return nil, fmt.Errorf("CSIPersistentVolumeSource not defined in spec")
+}
+
+func getCSIClient(volume, driverName string) *csiDriverClient {
```
Comment (vladimirvivien) [inline pkg/volume/csi/csi_util.go]: Again, please change the signature of this function to simply `getCSIClient(driverName string)`

File: pkg/volume/csi/csi_plugin.go [hunk 2]
```diff
@@ -31,26 +29,6 @@ import (
 	"k8s.io/kubernetes/pkg/volume"
 )
 
-const (
-	csiPluginName              = "kubernetes.io/csi"
-	csiVolAttribsAnnotationKey = "csi.volume.kubernetes.io/volume-attributes"
```
Comment (linyouchong) [inline pkg/volume/csi/csi_plugin.go]: I think it is better to leave `csiPluginName` and `csiVolAttribsAnnotationKey` where they are

File: pkg/volume/csi/csi_plugin.go [hunk 3]
```diff
@@ -198,20 +174,14 @@ func (p *csiPlugin) SupportsBulkVolumeVerification() bool {
...
-		glog.Error(log("unable to get kubernetes client from host"))
-		return nil, errors.New("unable to get Kubernetes client")
-	}
-
-	return &csiAttacher{
-		plugin:        p,
-		k8s:           k8s,
-		waitSleepTime: 1 * time.Second,
-	}, nil
+	return p.newCSIAttacher()
```
[the remark this answers is not in the document]
Comment (Lion-Wei) [reply pkg/volume/csi/csi_plugin.go]: I noticed `NewAttacher()` and `NewDetacher()` do exactly same thing, but can't directly use `NewAttacher()` in ` NewDetacher()`, cause they returned different interface. So I added a new method `newCSIAttacher()`.

File: pkg/volume/csi/csi_mounter.go [hunk 4]
```diff
@@ -263,10 +239,7 @@ func (c *csiMountMgr) TearDownAt(dir string) error {
 	driverName := data[volDataKey.driverName]
 
 	if c.csiClient == nil {
-		addr := fmt.Sprintf(csiAddrTemplate, driverName)
-		client := newCsiDriverClient("unix", addr)
-		glog.V(4).Infof(log("unmounter csiClient setup [volume=%v,driver=%v]", volID, driverName))
-		c.csiClient = client
+		c.csiClient = getCSIClient(volID, driverName)
```
[the remark this answers is not in the document]
Comment (Lion-Wei) [reply pkg/volume/csi/csi_mounter.go]: Thanks, already done. : )

File: pkg/volume/csi/csi_plugin.go (hunk 2 above)
Comment (Lion-Wei) [reply pkg/volume/csi/csi_plugin.go]: Thanks for the review, but may I know the reason?

File: pkg/volume/csi/csi_plugin.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (vladimirvivien) [reply pkg/volume/csi/csi_plugin.go]: Though `NewAttacher()` and `NewDetacher()` appears to do the same thing, but they are designed to be called in different situation (as they returned different types) and are not interchangeable.  There is no need to abstract them out.

[the remark this answers is not in the document]
Comment (Lion-Wei) [reply pkg/volume/csi/csi_plugin.go]: Oh, I see, thanks. : )

File: pkg/volume/csi/csi_plugin.go (hunk 2 above)
Comment (linyouchong) [reply pkg/volume/csi/csi_plugin.go]: Other Volume plugins do like this. Feel free to ignore this comment.

**Comment by vladimirvivien** [comment]
@Lion-Wei squash commits.

**Comment by Lion-Wei** [comment]
@vladimirvivien sorry for the delayed response, I was leave for Chinese New Year holiday.
Done, thanks. : )

**Comment by vladimirvivien** [comment]
@Lion-Wei no problem.  I will take a look.

File: pkg/volume/csi/csi_mounter.go [hunk 5]
```diff
@@ -288,48 +261,27 @@ func (c *csiMountMgr) TearDownAt(dir string) error {
...
-	dataFilePath := path.Join(dir, volDataFileName)
-
-	file, err := os.Create(dataFilePath)
-	if err != nil {
-		glog.Error(log("failed to save volume data file %s: %v", dataFilePath, err))
-		return err
+// getVolAttribsFromSpec extracts CSI VolumeAttributes information from PV.Annotations
+// using key csi.kubernetes.io/volume-attributes.  The annotation value is expected
+// to be a JSON-encoded object of form {"key0":"val0",...,"keyN":"valN"}
+func getVolAttribsFromSpec(spec *volume.Spec) (map[string]string, error) {
```
Comment (vladimirvivien) [inline pkg/volume/csi/csi_mounter.go]: this function can be completely removed as it is no longer used in the code.

File: pkg/volume/csi/csi_util.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/golang/glog"
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	kstrings "k8s.io/kubernetes/pkg/util/strings"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+const (
+	csiPluginName              = "kubernetes.io/csi"
+	csiVolAttribsAnnotationKey = "csi.volume.kubernetes.io/volume-attributes"
```
Comment (vladimirvivien) [inline pkg/volume/csi/csi_util.go]: delete csiVolAttribsAnnotationKey as it is no longer used.

File: pkg/volume/csi/csi_util.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+		spec.PersistentVolume.Spec.CSI != nil {
+		return spec.PersistentVolume.Spec.CSI, nil
+	}
+
+	return nil, fmt.Errorf("CSIPersistentVolumeSource not defined in spec")
+}
+
+func getCSIClient(driverName string) *csiDriverClient {
+	addr := fmt.Sprintf(csiAddrTemplate, driverName)
+	glog.V(4).Infof(log("unmounter csiClient setup driver=%v", driverName))
```
Comment (vladimirvivien) [inline pkg/volume/csi/csi_util.go]: Update the log message and remove the word "unmounter"

**Review by vladimirvivien — COMMENTED** [review]
Please see comments.

**Pull request closed without merging**